### PR TITLE
feat: local variables

### LINKED ISSUE #85: [definition] local variables

### PR DESCRIPTION
**Pull request by biletskyy**
resolves #85 

Add ability to go to local variable definition and references.

Correctly processes variables in 'cond' and defined with 'destructure'

### CODE REVIEW

**Comment by biletskyy** [comment]
Here is preview


https://github.com/elixir-tools/next-ls/assets/190730/ee4c822f-56be-47e5-af91-91faa9a1b8c8

File: lib/next_ls/helpers/ast_helpers.ex [hunk 1]
```diff
@@ -148,4 +148,228 @@ defmodule NextLS.ASTHelpers do
...
+      acc = Map.update!(acc, :sym_ranges, &[range | &1])
+      {ast, acc}
+    end
+
+    defp find_symbol(ast, acc), do: {ast, acc}
+
+    defp calculate_range(name, line, column) do
+      length = name |> to_string() |> String.length()
+
+      {line..line, column..(column + length)}
```
Comment (mhanberg) [inline lib/next_ls/helpers/ast_helpers.ex:318]: I think the column range might need a `- 1`

File: test/next_ls/helpers/ast_helpers_variables_test.exs [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+    end
+
+    test "symbol shadowed by anonymus funciton", %{source: source} do
+      refs = Variables.list_variable_references(source, {84, 12})
+      assert length(refs) == 2
+      assert exists_in?(refs, {:alpha, {96..96, 7..12}})
+      assert exists_in?(refs, {:alpha, {100..100, 21..26}})
+    end
+
+    test "symbol defined in anonymus funciton", %{source: source} do
```
Comment (mhanberg) [inline test/next_ls/helpers/ast_helpers_variables_test.exs:283]:
```suggestion
    test "symbol defined in anonymous funciton", %{source: source} do
```

File: test/next_ls/helpers/ast_helpers_variables_test.exs [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+      refs2 = Variables.list_variable_references(source, {74, 11})
+      assert length(refs2) == 1
+      assert exists_in?(refs2, {:delta, {75..75, 18..23}})
+
+      refs3 = Variables.list_variable_references(source, {78, 7})
+      assert length(refs3) == 1
+      assert exists_in?(refs3, {:error, {78..78, 25..30}})
+    end
+
+    test "symbol shadowed by anonymus funciton", %{source: source} do
```
Comment (mhanberg) [inline test/next_ls/helpers/ast_helpers_variables_test.exs:276]:
```suggestion
    test "symbol shadowed by anonymous funciton", %{source: source} do
```

File: test/next_ls/helpers/ast_helpers_variables_test.exs [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+      assert length(refs2) == 1
+      assert exists_in?(refs2, {:length, {127..127, 32..38}})
+
+      refs3 = Variables.list_variable_references(source, {123, 7})
+      assert length(refs3) == 1
+      assert exists_in?(refs3, {:var, {127..127, 17..20}})
+    end
+  end
+
+  defp exists_in?(results, var), do: Enum.find(results, &(&1 == var))
```
Comment (mhanberg) [inline test/next_ls/helpers/ast_helpers_variables_test.exs:337]: I think this reimplementing Enum.member?/2

Comment (mhanberg) [reply test/next_ls/helpers/ast_helpers_variables_test.exs:337]: Which also can be expressed as `assert item in results`

File: lib/next_ls/helpers/ast_helpers.ex [hunk 5]
```diff
@@ -148,4 +148,228 @@ defmodule NextLS.ASTHelpers do
...
+      {ast, acc}
+    end
+
+    # special case for 'cond', don't search for symbols in left side of 'cond' clause
+    defp prewalk({:->, meta, _} = ast, %{scope: ["cond" <> _ | _]} = acc) do
+      acc = increase_scope_nesting(acc, meta[:line])
+      {ast, acc}
+    end
+
+    # search symbols in a left side of forward arrow clause adn increase scope
```
Comment (mhanberg) [inline lib/next_ls/helpers/ast_helpers.ex:224]:
```suggestion
    # search symbols in a left side of forward arrow clause and increase scope
```

**Review by mhanberg — COMMENTED** [review]
Just a quick scan from my phone. Will need to look again later from the computer.

File: lib/next_ls/helpers/ast_helpers.ex (hunk 1 above)
Comment (biletskyy) [reply lib/next_ls/helpers/ast_helpers.ex:318]: yes, you are right, will fix

File: test/next_ls/helpers/ast_helpers_variables_test.exs (hunk 2 above)
Comment (biletskyy) [reply test/next_ls/helpers/ast_helpers_variables_test.exs:283]: sorry for spelling, should add spellchecker

File: test/next_ls/helpers/ast_helpers_variables_test.exs (hunk 4 above)
Comment (biletskyy) [reply test/next_ls/helpers/ast_helpers_variables_test.exs:337]: yes, you are totally right. Sorry, I just had a little brain fart. Will fix.

**Comment by mhanberg** [comment]
After we merge this, can you create tracking issues for the remaining TODOs?

**Comment by biletskyy** [comment]
> After we merge this, can you create tracking issues for the remaining TODOs?

Yes, I created issues #261 and #262. First one is probably liked to issue Can't find references of the contents of an inline function #184, as you mentioned before.

**Pull request merged**